design/ip-address-management-for-networkdata.md

```diff
@@ -0,0 +1,398 @@
+      prefix: 24
+    - start: 192.168.1.10
+      end: 192.168.1.15
+      subnet: 192.168.1.0/24
+      gateway: 192.168.1.1
+      prefix: 24
+  gateway: 192.168.1.1
+  prefix: 24
+  allocations:
+    "RenderedData-10": 192.168.0.9
+    "RenderedData-9": 192.168.0.8
+  namePrefix: "provisioning"
+status:
+  lastUpdated: "2020-04-02T06:36:09Z"
+  addresses:
+    "192.168.0.11": "RenderedData-1"
+  allocations:
+    "RenderedData-1": "pool-1-192-168-0-11"
+```
+
+In the *spec* of *IPPool*, there would be a *pools* list, that would
+contain a list of IP address pools, with the *start* and *end* attributes
+giving the start and end ip addresses of the pool. The *subnet* field allows to
+verify that the allocated IP is in the pool and from which the start and end ip
+addresses can be inferred. Specifying single ip addresses can be achieved by
+setting the start and end ip address to that single ip address.
+
+The *prefix* and *gateway* parameters can be given for each pool of the list,
+or globally. If they are given for a pool they will override the global
+settings, that are default values. The *prefix* and *gateway* will be set on
+the *IPAddress* and can be fetched from a *Template*.
+
+The *allocations* fields is a map of object name and ip address that allow a
+user to specify a set of static allocations for some objects.
+
+The *namePrefix* contains the prefix used to name the IPAddress objects
+created. It must remain the same for a subnet, across updates or changes in the
+IPPool object to keep the existing leases.
+
+The *status* would contain a *lastUpdated* field with the timestamp of the last
+update. In case of an error during the allocation (pool exhaustion for example),
+the error would be reported on the Consumer object, the *error* boolean and
+*errorMessage* field on the *RenderedData* object.
+The *allocations* map will map the IP address to the *RenderedData* object it was
+allocated for and the *addresses* will map the *RenderedData* objects with the
+*IPAddress* objects.
+
+The *IPAddress* object would be the following
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPAddress
+metadata:
+  name: pool-1-192-168-0-11
+  namespace: default
+  ownerReferences:
+  - apiVersion: ipam.metal3.io/v1alpha1
+    kind: IPPool
+    name: pool-1
+  - apiVersion: ipam.metal3.io/v1alpha1
+    kind: IPClaim
+    name: RenderedData-1
+spec:
+  ipClaim:
+    Name: RenderedData-1
+  Address: 192.168.0.11
+  prefix: 24
+  gateway: 192.168.0.1
+  IPPool:
+    Name: pool-1
+status:
+  ready: true
+```
+
+The *IPAddress* object would be the following
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPClaim
+metadata:
+  name: RenderedData-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: metadata.metal3.io/v1alpha1
+    kind: Data
+    name: data-1
+spec:
+  owner:
+    name: RenderedData-1
+  ipPool:
+    name: pool-1
+status:
+  ready: true
+  ipAddress:
+    name: pool-1-192-168-0-11
+  error: false
```

Comment:
If we don't need ready as a flag, I don't think we need error as a flag either.

Reply:
agreed!